vernac/himsg.ml

```diff
@@ -212,60 +212,78 @@ let explain_reference_variables sigma id c =
-      let pkp = Sorts.pr_sort_family kp in
-      let explanation =	match explanation with
-        | NonInformativeToInformative ->
-          "proofs can be eliminated only to build proofs"
-        | StrongEliminationOnNonSmallType ->
-          "strong elimination on non-small inductive types leads to paradoxes"
-        | WrongArity ->
-          "wrong arity" in
-      let ppar = match sorts with
-        | [] -> str "at some variable quality"
-        | _ -> pr_disjunction (fun s -> quote (Sorts.pr_sort_family s)) sorts
+    | None -> str "ill-formed elimination predicate."
+    | Some (pj, sp) ->
+      let ppt ?(ppunivs=false) () =
+        let pp () = pr_leconstr_env env sigma (mkSort (ESorts.make sp)) in
+        if ppunivs then Flags.with_option Constrextern.print_universes pp ()
+        else pp ()
       in
-      let ppt = pr_leconstr_env env sigma (snd (decompose_prod_decls sigma pj.uj_type)) in
-      hov 0
-        (str "the return type has sort" ++ spc () ++ ppt ++ spc () ++
-         str "while it" ++ spc () ++ str "should be " ++ ppar ++ str ".") ++
-      fnl () ++
-      hov 0
-        (str "Elimination of an inductive object of sort " ++
-         pki ++ brk(1,0) ++
-         str "is not allowed on a predicate in sort " ++ pkp ++ fnl () ++
-         str "because" ++ spc () ++ str explanation ++ str ".")
-  | None ->
-      str "ill-formed elimination predicate."
+      let squash = Option.get (Inductive.is_squashed (specif, snd ind)) in
+      match squash with
+      | SquashToSet ->
+        let ppt = ppt () in
+        hov 0
+          (str "the return type has sort" ++ spc () ++ ppt ++ spc () ++
+           str "while it should be SProp, Prop or Set.") ++
+        fnl () ++
+        hov 0
+          (str "Elimination of an inductive object of sort Set" ++ spc() ++
+           str "is not allowed on a predicate in sort " ++ ppt ++ fnl () ++
+           str "because" ++ spc () ++
+           str "strong elimination on non-small inductive types leads to paradoxes.")
+      | SquashToQuality (QConstant (QSProp | QProp as squashq)) ->
+        let ppt = ppt () in
+        let inds, sorts = match squashq with
+          | QSProp -> "SProp", "SProp"
+          | QProp -> "Prop", "SProp or Prop"
+          | QType -> assert false
+        in
+        hov 0
+          (str "the return type has sort" ++ spc () ++ ppt ++ spc () ++
+           str "while it should be " ++ str sorts ++ str ".") ++
+        fnl () ++
+        hov 0
+          (str "Elimination of an inductive object of sort " ++ str inds ++ spc() ++
+           str "is not allowed on a predicate in sort " ++ ppt ++ fnl () ++
+           str "because" ++ spc () ++
+           (* we could try to be more precise by adding "strict" qualifiers here
+              but IMO not worth making this text more complicated when we already
+              say which of SProp/Prop work just above. *)
+           str "proofs can be eliminated only to build proofs.")
```

Comment:
actually we should probably have something in the case where the inductive is SProp and the predicate is Prop